src/PooledArrays.jl

```diff
@@ -304,7 +307,64 @@ Base.findall(pdv::PooledVector{Bool}) = findall(convert(Vector{Bool}, pdv))
 ##
 ##############################################################################
 
-function Base.map(f, x::PooledArray{T,R}) where {T,R<:Integer}
+"""
+    map(f, x::PooledArray; pure::Bool=false)
+
+Transform `PooledArray` `x` by applying `f` to each element.
+
+If `pure=true` then `f` is applied to each element of pool of `x`
+exactly once (even if some elements in pool are not present it `x`).
+This will typically be much faster when the proportion of unique values
+in `x` is small.
+
+If `pure=false` the value returned by `map` is not type stable.
+"""
+function Base.map(f, x::PooledArray{<:Any, R, N, RA}; pure::Bool=false)::Union{PooledArray{<:Any, R, N, RA},
+                                                                               PooledArray{<:Any, Int64, N, Vector{Int64}}} where {R, N, RA}
```

Comment:
Unfortunately `Vector{Int64}` isn't correct in general as we call `labels = similar(x.refs)` and then `convert(AbstractArray{I2}, labels)`. I'm not sure how to compute the right type, except using `typeof(similar(x.refs, Int64, 0))`, which is quite wasteful.

Reply:
ah - I keep forgetting about it. Actually the correct way is `typeof(similar(x.refs, Int64, ntuple(i -> 0, ndims(x.refs))))` I think.